server/src/main/java/io/horizen/services/impl/RosettaServiceImpl.java

```diff
@@ -0,0 +1,185 @@
+
+        Helper.warnIfProposalNotActive(snapshotService.getActiveProposal());
+
+        Map<String, List<String>> mcAddressMap;
+        try {
+            mcAddressMap = snapshotService.getMcAddressMap(scAddress);
+        }
+        catch (Exception ex) {
+            throw new GetMcAddressMapException();
+        }
+        if (mcAddressMap.containsKey(scAddress)) {
+            List<String> mcAddresses = mcAddressMap.get(scAddress);
+
+            if (settings.getMockRosetta()) {
+                System.out.println("MOCK ROSETTA RESPONSE");
+                return 123456789.0;
+            }
+
+            int blockHeight = snapshotService.getActiveProposal().getBlockHeight();
+            String blockHash = snapshotService.getActiveProposal().getBlockHash();
+
+            String actualBlockHash = getMainchainBlockHash(blockHeight);
+            if (!actualBlockHash.equals(blockHash)) {
+                log.info("MC block hash mismatch. Using hash=" + actualBlockHash + " (instead of " + blockHash + ") for height=" + blockHeight);
+                blockHash = actualBlockHash;
+            }
+
+            for (String mcAddress : mcAddresses) {
+                String body = buildRosettaRequestBodyForNetworkStatus(mcAddress, blockHash);
+                HttpURLConnection connection = Helper.sendRequest(settings.getRosettaUrl() + "account/balance", body);
+
+                if(connection.getResponseCode() == HttpURLConnection.HTTP_OK) {
+                    String response = IOUtils.toString(connection.getInputStream());
+                    JsonElement jsonElement = JsonParser.parseString(response);
+
+                    int amount = jsonElement
+                            .getAsJsonObject()
+                            .getAsJsonArray("balances")
+                            .get(0)
+                            .getAsJsonObject()
+                            .get("value")
+                            .getAsInt();
+                    balance += amount;
+                }
+            }
+        }
+        return balance;
+    }
+
+
+    private String buildRosettaRequestBodyForNetworkStatus(String mcAddress, String blockHash) {
+        JsonObject rosettaRequestTemplate = new JsonObject();
+        rosettaRequestTemplate.addProperty("blockchain", "Zen");
+        rosettaRequestTemplate.addProperty("network", settings.getNetwork());
+
+        JsonObject requestBody = new JsonObject();
+        requestBody.add("network_identifier", rosettaRequestTemplate);
+
+        JsonObject accountIdentifier = new JsonObject();
+        accountIdentifier.addProperty("address", mcAddress);
+        accountIdentifier.add("metadata", new JsonObject());
+        requestBody.add("account_identifier", accountIdentifier);
+
+        JsonObject blockIdentifier = new JsonObject();
+        blockIdentifier.addProperty("index", snapshotService.getActiveProposal().getBlockHeight());
+        blockIdentifier.addProperty("hash", blockHash);
```

Comment:
We can avoid using hash here and keeping only height. The reason is that this block might have been reverted in mainchain since we first get it upon proposal creation.

Reply:
Done